components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/graphql/GraphQLOperationAnalyzer.java

```diff
@@ -0,0 +1,558 @@
+    private static final String MUTATION = "mutation";
+    private static final String SUBSCRIPTION = "subscription";
+
+    /**
+     * @param type Operation Type
+     * @param selection Selection Set
+     * @param subDocument Parsed document
+     * @param graphQLSchema GraphQL Schema
+     * @param typeDefinitionRegistry GraphQL Type Definition Registry
+     * @param complexityInfoJson Complexity Info
+     * @param variableMap External Variables Map
+     * @return field usages, depth, complexities of the requested query
+     */
+    public static Map<String, Object> getOperationInfo(String type, Selection selection, Document subDocument,
+                                                       GraphQLSchema graphQLSchema,
+                                                       TypeDefinitionRegistry typeDefinitionRegistry,
+                                                       String complexityInfoJson, Map<String, Object> variableMap) {
+
+        Map<String, Object> operationInfo = new HashMap<>();
+        operationInfo.put(OPERATION_DEPTH, calculateDepth(subDocument, graphQLSchema, variableMap));
+        operationInfo.put(OPERATION_COMPLEXITY, calculateComplexity(subDocument, graphQLSchema,
+                complexityInfoJson, variableMap));
+
+        //add query name to custom properties
+        String operationName;
+        if (selection instanceof Field) {
+            Field field = (Field) selection;
+            operationName = field.getName();
+            operationInfo.put(OPERATION_NAME, operationName);
+        } else {
+            throw new IllegalArgumentException("Expected a Field but got a different type of Selection");
+        }
+
+        //add return type to custom properties
+        GraphQLObjectType objectType = null;
+        if (type.equals(QUERY)) {
+            objectType = graphQLSchema.getQueryType();
+        } else if (type.equals(MUTATION)) {
+            objectType = graphQLSchema.getMutationType();
+        } else if (type.equals(SUBSCRIPTION)) {
+            objectType = graphQLSchema.getSubscriptionType();
+        }
+
+        if (objectType != null) {
+            String returnType = getOperationReturnType(objectType, operationName);
+            operationInfo.put(RETURN_TYPE, returnType);
+        }
+
+        // Add all used types in a given query
+        List<String> parentList = new ArrayList<>();
+        for (Definition subDefinition : subDocument.getDefinitions()) {
+            OperationDefinition subOperation =  (OperationDefinition) subDefinition;
+            Map<String, Object> typesUsedByEachOperation = getTypeList(subOperation, typeDefinitionRegistry);
+            List<Map<String, String>> usedList = (List<Map<String, String>>) typesUsedByEachOperation.get("used");
+            for (Map<String, String> entry : usedList) {
+                String parent = entry.get("parent");
+                if (!parentList.contains(parent)) {
+                    parentList.add(parent);
```

Comment:
Why do we need to add parent value into parentList here? Can't we fill the accessedTypesList here it self rather than iterating parentList in L143

Reply:
Fixed via https://github.com/wso2/carbon-apimgt/pull/12441/commits/226c49a261240dc5305f96d69aadb6f86edfae84